README.md
# api-docs-gen

> api docs generator


## Dev

```
$ npm install && bower install
```

### Run

```
$ npm start
```

### Build

```
$ npm run wizard
```

Builds the app for OS X, Linux, and Windows, using [electron-packager](https://github.com/electron-userland/electron-packager).


## License

MIT © [](https://github.com/Linko91)


CHANGELOG.md
CHANGELOG
====

`1.2.0` 02/08/2016 11:08
----
 * update changelog.
 * update icon.
 * update changelog.
 * add icon.

`1.1.0` 01/08/2016 17:52
----
 * update changelog.
 * add build command.
 * update export features (search).
 * add  search feature.

`1.0.0` 28/06/2016 10:05
----
 * update changelog.
 * update export layout.
 * update export layout.
 * update changelog.

`0.12.0` 27/06/2016 13:25
----
 * update version.
 * add autosave function.
 * add unsave control.
 * update autosave slider.
 * add autosave layout.
 * update export function.
 * update export sidebar layout.
 * update sidebar layout.
 * update export layout.
 * update save flow.

`0.11.1` 23/06/2016 18:11
----
 * update save and export functions.
 * update export-res.
 * update dependencies.
 * update label "path" to "endpoint".

`0.11.0` 01/06/2016 18:29
----
 * update changelog.
 * add export feature.
 * remove scroll log.

`0.10.1` 31/05/2016 21:07
----
 * fix sticky header.
 * add new custom sticky plugin.
 * update sticky plugin.
 * update bower resolutions.
 * add angular-sticky plugin.
 * update form format.
 * update version.
 * update adg format.
 * fix adg split.
 * add resize box to ace.
 * fix resize controll.

`0.10.0` 29/05/2016 16:51
----
 * add ace editor.
 * update "formTest".

`0.9.0` 27/05/2016 17:36
----
 * add responses management.
 * update changelog.
 * add query & headers management.
 * update version.
 * add Query string layout.

`0.8.0` 26/05/2016 18:28
----
 * fix service-box animation.
 * fix os-mac icon css.
 * update dependencies.
 * fix mac os animations.
 * add os detector.
 * add animations.
 * update sidebar theme.
 * update theme.
 * update theme.
 * add services management.

`0.7.0` 23/05/2016 23:12
----
 * add bases and paths management.
 * add new modal component.
 * update dependencies.
 * add dialog template.
 * update changelog.

`0.6.0` 19/05/2016 18:26
----
 * update version -> 0.6.0.
 * add "open & save" file.

`0.5.1` 18/05/2016 18:25
----
 * implement navbar ng-repeat.
 * update actions buttons.
 * add changelog.

`0.5.0` 18/05/2016 10:50
----
 * add navbar -> list-view.

`0.4.1` 17/05/2016 20:15
----
 * update bower dependencies.

`0.4.0` 17/05/2016 17:57
----
 * add navbar-resizable view.
 * sidebar layout.

`0.3.0` 16/05/2016 23:11
----
 * add "edit api" page.

`0.2.0` 16/05/2016 19:14
----
 * add "new api" page.

`0.1.2` 16/05/2016 17:55
----
 * implemented drag & drop.
 * add drag & drop module.
 * increased font compatibility.

`0.1.1` 16/05/2016 11:57
----
 * add bower dependencies.
 * update window options.
 * update readme.

`0.1.0` 16/05/2016 09:58
----
 * initial commit.
 * :lollipop: Added .gitattributes.


index.js
'use strict';
const electron = require('electron');
const Menu = electron.Menu;
const MenuItem = electron.MenuItem;
const app = electron.app;

// adds debug features like hotkeys for triggering dev tools and reload
require('electron-debug')({showDevTools: 'undocked'});

// prevent window being garbage collected
let mainWindow;

function onClosed() {
	// dereference the window
	// for multiple windows store them in an array
	mainWindow = null;
}

function createMainWindow() {
	const win = new electron.BrowserWindow({
		width: 1000,//600
		height: 800,//400
		darkTheme: true,
		title: "Api Docs Generator",
		directWrite:true,
    icon: __dirname + '/icon.png'
		//transparent: true, frame: false
	});

	win.loadURL('file://' + __dirname + '/index.html');
	win.on('closed', onClosed);

	return win;
}

app.on('window-all-closed', () => {
	if (process.platform !== 'darwin') {
		app.quit();
	}
});

app.on('activate', () => {
	if (!mainWindow) {
		mainWindow = createMainWindow();
	}
});

app.on('ready', () => {
	mainWindow = createMainWindow();
});




/*
var template = [
  {
    label: 'Edit',
    submenu: [
      {
        label: 'Cut',
        accelerator: 'CmdOrCtrl+X',
        role: 'cut'
      },
      {
        label: 'Copy',
        accelerator: 'CmdOrCtrl+C',
        role: 'copy'
      },
      {
        label: 'Paste',
        accelerator: 'CmdOrCtrl+V',
        role: 'paste'
      },
      {
        label: 'Select All',
        accelerator: 'CmdOrCtrl+A',
        role: 'selectall'
      },
    ]
  },
  {
    label: 'View',
    submenu: [
      {
        label: 'Toggle Full Screen',
        accelerator: (function() {
          if (process.platform == 'darwin')
            return 'Ctrl+Command+F';
          else
            return 'F11';
        })(),
        click: function(item, focusedWindow) {
          if (focusedWindow)
            focusedWindow.setFullScreen(!focusedWindow.isFullScreen());
        }
      }
    ]
  },
  {
    label: 'Window',
    role: 'window',
    submenu: [
      {
        label: 'Minimize',
        accelerator: 'CmdOrCtrl+M',
        role: 'minimize'
      },
      {
        label: 'Close',
        accelerator: 'CmdOrCtrl+W',
        role: 'close'
      },
    ]
  },
  {
    label: 'Help',
    role: 'help',
    submenu: [
      {
        label: 'Learn More',
        click: function() { electron.shell.openExternal('http://electron.atom.io') }
      },
    ]
  },
];

if (process.platform == 'darwin') {
  var name = electron.app.getName();
  template.unshift({
    label: name,
    submenu: [
      {
        label: 'About ' + name,
        role: 'about'
      },
      {
        type: 'separator'
      },
      {
        label: 'Services',
        role: 'services',
        submenu: []
      },
      {
        type: 'separator'
      },
      {
        label: 'Hide ' + name,
        accelerator: 'Command+H',
        role: 'hide'
      },
      {
        label: 'Hide Others',
        accelerator: 'Command+Alt+H',
        role: 'hideothers'
      },
      {
        label: 'Show All',
        role: 'unhide'
      },
      {
        type: 'separator'
      },
      {
        label: 'Quit',
        accelerator: 'Command+Q',
        click: function() { app.quit(); }
      },
    ]
  });
  // Window menu.
  template[3].submenu.push(
    {
      type: 'separator'
    },
    {
      label: 'Bring All to Front',
      role: 'front'
    }
  );
}

var menu = Menu.buildFromTemplate(template);
Menu.setApplicationMenu(menu);
*/

assets/js/app.js
/*
 _        _         _            
| \      (_)       | |           
| |       _  ____  | |  _   ___  
| |      | ||  _ \ | |_/ ) / _ \ 
| |_____ | || | | ||  _ ( | (_) |
\_______)|_||_| |_||_| \_) \___/ 

*/

(function(){ // I <3 Closure
'use strict';

var DEBUG = true;


const Config = require('electron-config');
const config = new Config();
const {dialog} = require('electron').remote;
var fs = require('fs'); // Load the File System to execute our common tasks (CRUD)

require('electron-context-menu')();


var app = angular.module('app', ['ngMaterial', 'ngAnimate', 'ngFileUpload', 'ngAria', 'ngMessages', 'angularResizable', 'ng.deviceDetector', 'ui.ace']);

app.constant('DEBUG', DEBUG);
app.constant('dropzoneclass', 'drop-box');
app.constant('filext', 'adg');

app.config(function($mdThemingProvider) {
	/*var asphalt = $mdThemingProvider.extendPalette('blue', {
	    '500': '17395F',
	    '300': '295280',
	    '800': '112133',
	  });
	$mdThemingProvider.definePalette('asphalt', asphalt);
	$mdThemingProvider.theme('default').primaryPalette('asphalt');*/
	
	$mdThemingProvider.theme('default')
	    .primaryPalette('blue-grey')
	    .accentPalette('teal')
	    .warnPalette('red')
	    .backgroundPalette('blue-grey')
	    .dark();

    _.mixin(s.exports());
    moment.locale('it');
});

app.controller('appCtrl', function($scope, $mdDialog, $sce, $mdToast, $interval, $http, DEBUG, dropzoneclass, filext, $timeout, $mdSidenav, $log, $mdMedia, deviceDetector){
	var vm = this;
	/////////////////////////////////////
	////////////// CTRL VARS ////////////
	/////////////////////////////////////

	if(config.get('autosave') === undefined)
		config.set('autosave', true);

	if(config.get('autotime') === undefined)
		config.set('autotime', 5);

	vm.config = {
		unsaved: false,
		saving: false,
		autosave: config.get('autosave'),
		autotime: config.get('autotime') //minutes
	};

	var timingsave;

	vm.page = {
		initpage : 1,
		newapi   : 0,
		editapi  : 0
	};

	vm.subpage = {
		editpro  : 0,
		editsrv  : 0,
		editbase : 0
	};

	vm.deviceDetector = deviceDetector;

	vm.navbarClass = 2;// 3=ico+text   2=text   1=ico

	vm.content_type_req = ['text', 'xml', 'json', 'form'];
	vm.content_type_res = ['text', 'xml', 'json'];
	vm.method_list = ['GET', 'POST', 'PUT', 'DELETE', 'OPTIONS', 'PATCH', 'TRACE', 'HEAD'];

	vm.filePath = '';
	vm.fileObj = '';
	vm.formTmp = {};
	vm.formIdsTmp = [];

	vm.formDefault = {
		project: {
			name          : '',
			description   : '',
			mime_request  : 'json',
			mime_response : 'json'
		},
		bases: []
	};

	vm.formTest = {
		project: {
			name          : 'Name of project',
			description   : 'project description',
			mime_request  : 'json',
			mime_response : 'json'
		},
		bases: [
			{
				name: 'http://test.io/service',
				description: '',
				paths: [
					{
						name:'/action',
						description: '',
						services: [
							{
								path:'/action',
								name:'Manage actions',
								description: '',
								method: 'POST',
								headers: [
									{name:'token', desc:'Id token'},
									{name:'user', desc:'id user'}
								],
								query: [
									{name:'namequno', desc:'descquno'},
									{name:'nameqdue', desc:'descqdue'}
								],
								form: [
									{name:'namefuno', desc:'descfuno'},
									{name:'namefdue', desc:'descfdue'}
								],
								body: 'test',
								type: 'json',
								responses: [
									{code:'200', body:'', type:'json', desc:''},
									{code:'404', body:'', type:'json', desc:''},
									{code:'405', body:'', type:'json', desc:''}
								]
							}
						]
					}
				]
			}
		]
	};

	vm.form = angular.copy(vm.formDefault);

	//form
	//{name:'nameuno', type:'typeuno', desc:'descuno'}

	//json, xml, text
	//	free - example json:
	//		{
	//			id: 'int',
	//			name: 'string - optional',
	//			slug: 'string',
	//			tags: 'array'
	//		}
			
	



	/////////////////////////////////////
	//////////// CTRL METHOD ////////////
	/////////////////////////////////////
	vm.init = function(){
		vm.dropzonevents();
	};

	vm.dropzonevents = function (){
		window.addEventListener('dragenter', function (event) {
			return vm.checkDropzone(event);
		});
		window.addEventListener('dragover', function (event) {
			return vm.checkDropzone(event);
		});
		window.addEventListener('dragleave', function (event) {
			return vm.checkDropzone(event);
		});
		window.addEventListener('drop', function (event) {
			return vm.checkDropzone(event);
		});		
	};

	vm.checkDropzone = function(e){
		if(DEBUG){console.log(e.dataTransfer.files)};
		var ext = '-';
		if(e.dataTransfer.files && e.dataTransfer.files.length && e.dataTransfer.files[0].name && e.dataTransfer.files[0].name.indexOf('.') != -1){
			ext = e.dataTransfer.files[0].name.split('.').pop();
		}

		if (e.target.className.indexOf(dropzoneclass) === -1 || e.dataTransfer.files.length > 1 || ext != filext) {
		//if (e.target.className.indexOf(dropzoneclass) === -1) {
			e.preventDefault();
			e.dataTransfer.effectAllowed = "none";
			e.dataTransfer.dropEffect = "none";
			return false;
		}else{
			return true;
		}
	};

	vm.switchPage = function(page, sub){
		var type = 'page';
		if(sub && sub === 1){
			type = 'subpage';
		}

		if(!vm[type][page]){
			for(var key in vm[type]){
				vm[type][key] = 0;
			}

			if(page === 'editapi' && type === 'page'){
				vm.resetEditPage();
			}

			vm[type][page] = 1;
		}
	};

	vm.resetEditPage = function(){
		vm.navbarClass = 2;
		vm.subpage.editpro = 0;
		vm.subpage.editsrv = 0;
		vm.subpage.editbp = 0;
	};

	vm.gotoHome = function(){
		vm.formTmp = {};
		vm.filePath = '';
		vm.form = angular.copy(vm.formDefault);
		vm.fileObj = angular.copy(vm.formDefault);
		vm.config.unsaved = false;
       	vm.config.saving = false;
       	vm.timingAutoSave();
		vm.switchPage('initpage');
	}

	vm.uploadADG = function(file) {
		if(DEBUG){console.log(file);}
		if(file) vm.getDataUri(file);
	};

	vm.loadADG = function(data, file) {
		if(DEBUG){console.log(data);}
		var error = 0;

		if(data && data.indexOf('@end@') != -1){
			try {
			    var arr = data.split('@end@');
			    var json = JSON.parse(_.rest(arr).join('@end@'));
			    vm.fileObj = angular.copy(json);
			    vm.form = angular.copy(json);
			    
			    vm.filePath = file.path;
	        	vm.switchPage('editapi');
	        	vm.timingAutoSave();
			}
			catch(err) {
			    if(DEBUG){console.log(err);}
			    error = 1;
			}
		}else{
			error = 1;
		}

		if(error){
			$mdDialog.show(
		      $mdDialog.alert()
		        //.parent(angular.element(document.querySelector('#popupContainer')))
		        .clickOutsideToClose(true)
		        .title('ERROR')
		        .textContent('the file is corrupted')
		        .ariaLabel('file corrupted')
		        .ok('OK')
		        //.targetEvent(ev)
		    );
		}
	};

	vm.getDataUri = function(file) {
        var reader = new FileReader();

        // Closure to capture the file information.
        reader.onload = (function(theFile) {
            return function(e) {
                if(DEBUG){console.log(e, e.target.result);}
                vm.loadADG(e.target.result, theFile);
            };
        })(file);

        // Read in the file as a text.
        reader.readAsText(file);
    };

    vm.saveADG = function() {
    	if(DEBUG){console.log('saving')};

    	vm.config.saving = true;
    	var filename = _.slugify(vm.form.project.name) || 'data';
		var obj = angular.copy(vm.form);
		/*var data = "text/plain;charset=utf-8," + encodeURIComponent("#api-docs-gen@end@"+JSON.stringify(obj));
		var a = document.createElement('a');
		a.download = filename+'.adg';
		a.href = 'data:' + data;
		a.click();*/

		if(vm.filePath){
	        // fileName is a string that contains the path and filename created in the save file dialog.  
	        fs.writeFile(vm.filePath, "#api-docs-gen@end@"+JSON.stringify(obj), function (err) {
	            if(err){
	                alert("An error ocurred creating the file "+ err.message)
	            }else{
		           	vm.fileObj = angular.copy(obj);
		           	vm.config.unsaved = false;
		           	vm.config.saving = false;
		           	if(DEBUG){console.log('saved')};
	            }
	                 
	            //alert("The file has been succesfully saved");
	        });
		}else{
			dialog.showSaveDialog({title:'Save project', defaultPath:filename+'.adg', buttonLabel:'Save'}, function (fileName) {
		        if (fileName === undefined){
		            if(DEBUG){console.log("You didn't save the file");}
		            vm.config.saving = false;
		            return;
		        }
		        // fileName is a string that contains the path and filename created in the save file dialog.  
		        fs.writeFile(fileName, "#api-docs-gen@end@"+JSON.stringify(obj), function (err) {
		            if(err){
		                alert("An error ocurred creating the file "+ err.message)
		            }else{
			           	vm.fileObj = angular.copy(obj);
			           	vm.config.unsaved = false;
			           	vm.config.saving = false;
			            vm.filePath = fileName;
			            alert("The file has been succesfully saved");
		            }
		        });
			}); 			
		}

    };

    vm.export = function() {
    	var filename = _.slugify(vm.form.project.name) || 'data';
		var obj = angular.copy(vm.form);
		//console.log(obj);

		for(var k_base in obj.bases){
			for(var k_path in obj.bases[k_base].paths){
				for(var k_service in obj.bases[k_base].paths[k_path].services){
					obj.bases[k_base].paths[k_path].services[k_service].isOpen = false;
				}
			}
		}

		var html = getHtml(filename, obj, 0);

		/*var data = "text/html;charset=utf-8," + encodeURIComponent(html);
		var a = document.createElement('a');
		a.download = filename+'.html';
		a.href = 'data:' + data;
		a.click();*/

		dialog.showSaveDialog({title:'Export Api Doc', defaultPath:filename+'.html', buttonLabel:'Export'}, function (fileName) {
	       if (fileName === undefined){
	            if(DEBUG){console.log("You didn't save the file");}
	            return;
	       }
	       // fileName is a string that contains the path and filename created in the save file dialog.  
	       fs.writeFile(fileName, html, function (err) {
	           if(err){
	               alert("An error ocurred creating the file "+ err.message)
	           }
	                        
	           alert("The file has been succesfully saved");
	       });
		}); 
    };

    vm.timingAutoSave = function(){
    	if(vm.filePath){
			if(vm.config.autosave){
				timingsave = $interval(vm.saveADG, vm.config.autotime*60*1000);
			}else{
				$interval.cancel(timingsave);	
			}
		}else{
			$interval.cancel(timingsave);	
		}
    }

	/**
	 * Editing base & path
	 * @param {Number} act  (1=new, 2=edit, 3=del) 
	 * @param {Number} type (1=base, 2=path)
	 * @param {Object} obj  (base or path)
	 * @param {Array}  ids  ([id base, id path] or [id base] or [])
	 */
	vm.editBasePath = function(act, type, obj, ids) { 
		if(act === 1){ //new
			if(type === 1){
		    	vm.form.bases.push({name: '', description: '', paths: []});
			}else if(type === 2){
				if(obj){
			    	obj.paths.push({name: '', description: '', services: []});
			    }
			}
		}else if(act === 2){ //edit
			if(obj){
				vm.formIdsTmp = ids;			
				vm.formTmp = obj;
				if(type === 1){
			    	vm.switchPage('editbase', 1);
				}else if(type === 2){
			    	vm.switchPage('editsrv', 1);
				}
			}
		}else if(act === 3){ //del
			if(vm.formIdsTmp.length === 1){	//base
				vm.form.bases.splice(vm.formIdsTmp[0], 1);     
				vm.subpage.editbase = 0;
				//delete vm.form.bases[vm.formIdsTmp[0]];
				//_.compact(vm.form.bases);
			}else if(vm.formIdsTmp.length === 2){ //path
				vm.form.bases[vm.formIdsTmp[0]].paths.splice(vm.formIdsTmp[1], 1);     
				vm.subpage.editsrv = 0;
				//delete vm.form.bases[vm.formIdsTmp[0]].paths[vm.formIdsTmp[1]];
				//_.compact(vm.form.bases[vm.formIdsTmp[0]].paths);
			}		
		}
	};


	/**
	 * Editing service
	 * @param {Number} act  (1=new, 2=del) 
	 * @param {Object} obj  (path)
	 * @param {Number}  id  (id service)
	 */
	vm.editService = function(act, obj, id){
		if(act === 1){ //new
			var type = angular.copy(vm.form.project.mime_request);
			var name = angular.copy(obj.name);
			obj.services.push({
				path:name,
				name:'',
				description: '',
				method: '',
				headers: [],
				query: [],
				form: [],
				body: '',
				type: type,
				responses: []
			});
		}else if(act === 2){ //del
			obj.services.splice(id, 1);     
		}
	};


	/**
	 * Editing query string & headers
	 * @param {Number} act  (1=new, 2=del) 
	 * @param {Number} type (1=query, 2=header, 3=form, 4=response)
	 * @param {Object} obj  (service)
	 * @param {Number} id   (id query or id header)
	 */
	vm.editQueryHeaderForm = function(act, type, obj, id){
		var key = '';

		switch(type){
			case 1:
				key = 'query';
				break;			
			case 2:
				key = 'headers';
				break;
			case 3:
				key = 'form';
				break;
			case 4:
				key = 'responses';
				break;
		}

		if(act === 1 && key){ //new
			if(type === 4){
				var type = angular.copy(vm.form.project.mime_response);
				obj[key].push({code:'', body:'', type:type, desc:''});
			}else{
				obj[key].push({name: '', desc: ''});
			}
		}else if(act === 2 && key){ //del
			obj[key].splice(id, 1);     
		}
	};


	/////////////////////////////////////
	//////////// INIT METHOD ////////////
	/////////////////////////////////////
	vm.init();
	vm.timingAutoSave();






	/////////////////////////////////////
	///////////// LISTENERS /////////////
	/////////////////////////////////////
	/*$scope.$on("angular-resizable.resizeEnd", function (event, args) {
		console.log('angular-resizable.resizeEnd', args);
    });
	$scope.$on("angular-resizable.resizeStart", function (event, args) {
		console.log('angular-resizable.resizeStart', args);
    });*/
	$scope.$on("angular-resizable.resizing", function (event, args) {
		if(DEBUG){console.log('angular-resizable.resizing', args);}

		if(args.id === 'navbar'){
			if(args.width >= 300){
				vm.navbarClass = 3;
			}else if(args.width < 300 && args.width >= 210){
				vm.navbarClass = 2;
			}else{
				vm.navbarClass = 1;
			}
		}
    });


	$scope.$watch("vm.form", function (newValue, oldValue) {
		//if (newValue != oldValue)
	        vm.config.unsaved = !angular.equals(newValue, vm.fileObj);
    }, true);

	$scope.$watch("vm.config.autosave", function (newValue, oldValue) {
		if (newValue != oldValue){
	        config.set('autosave', newValue);

        	vm.timingAutoSave();
		}
    });

	$scope.$watch("vm.config.autotime", function (newValue, oldValue) {
		if (newValue != oldValue){
	        config.set('autotime', newValue);

	        vm.timingAutoSave();
		}
    });





}); //end appCtrl

app.directive('stickyContained', ['$window', function($window) {
	var stickies = [];

	var scroll = function scroll() {
		//console.log(arguments);

		for(var i=0; i<stickies.length; i++){

			var containerOffset = stickies[i].container.offset();
			var offsetTop = containerOffset.top;
			var elHeight = stickies[i].el.outerHeight();
			var elOffset = stickies[i].el.offset();
			var elPos = stickies[i].el.position();
			var containerHeight = stickies[i].container.outerHeight();
			var containerVis = containerHeight-Math.abs(offsetTop);

			if(offsetTop < -stickies[i].position.top){
				
				stickies[i].el.addClass('isSticky');
			
				//console.log(containerVis, elHeight, containerHeight, offsetTop);
				if(containerVis<=elHeight){
					stickies[i].el.css('top', Math.abs(containerHeight-elHeight));
				}else{
					stickies[i].el.css('top', Math.abs(offsetTop));
				}						

			}else{
				stickies[i].el.css('top', stickies[i].position.top).removeClass('isSticky');
			}
			
			//console.log('scroll', stickies[i].el.text(), containerHeight-Math.abs(offsetTop), stickies[i].container, stickies);
		}
	};
	var link = function($scope, element, attrs) {
		//console.log(arguments);

		var scroller = angular.element($window);
		var container = angular.element($window);
		var el = element;

		if(attrs.scrollContainer)
			container = scroller = element.closest(attrs.scrollContainer);

		if(attrs.stickyContainer)
			container = element.closest(attrs.stickyContainer);

		if(DEBUG){console.log(el.position())};

	    angular.element(scroller).on('scroll', scroll);

		stickies.push({scroller:scroller, container:container, el:el, position:el.position()});

		$scope.$on('$destroy', function() {
	        //console.log("destroy", _.findIndex(stickies, {el: element}));

	        var i = _.findIndex(stickies, {el: element});
	        angular.element(stickies[i].scroller).off('scroll', scroll).on('scroll', scroll);
	        stickies.splice(i, 1);
	    });

	};
    
    return {
		restrict: 'A',
		link: link
    };
}]); //end directive


})(); //end closure

